fluent-react/src/provider.js

```diff
@@ -21,57 +23,80 @@ import createParseMarkup from "./markup";
  * `ReactLocalization` to format translations.  If a translation is missing in
  * one instance, `ReactLocalization` will fall back to the next one.
  */
-export default class LocalizationProvider extends Component {
-  constructor(props) {
-    super(props);
-    const {bundles, parseMarkup} = props;
-
-    if (bundles === undefined) {
-      throw new Error("LocalizationProvider must receive the bundles prop.");
-    }
-
-    if (!bundles[Symbol.iterator]) {
-      throw new Error("The bundles prop must be an iterable.");
-    }
-
-    this.l10n = new ReactLocalization(bundles);
-    this.parseMarkup = parseMarkup || createParseMarkup();
+export default function LocalizationProvider(props) {
+  if (props.bundles === undefined) {
+    throw new Error("LocalizationProvider must receive the bundles prop.");
   }
 
-  getChildContext() {
-    return {
-      l10n: this.l10n,
-      parseMarkup: this.parseMarkup,
-    };
+  if (!props.bundles[Symbol.iterator]) {
+    throw new Error("The bundles prop must be an iterable.");
   }
 
-  componentWillReceiveProps(next) {
-    const { bundles } = next;
+  const bundles = useMemo(() => CachedSyncIterable.from(props.bundles), [
+    props.bundles
+  ]);
+  const parseMarkup = useMemo(() => props.parseMarkup || createParseMarkup(), [
+    props.parseMarkup
+  ]);
+  const contextValue = useMemo(
+    () => {
+      const l10n = {
```

Comment:
This would be in line with your proposal from Slack to memoize the entire context rather than individual props, right?

Reply:
I wanted to (and now did) memoize the whole Provider component and not just the contextValue. That should make things a lot clearer. I also added a test to make sure that it stop unnecessary re-renders. It doesn't have the drawbacks of `useMemo`, namely
- no extra per-render array+function-scope allocations
- (probably) no future react might forget your memoization (at least not according to the docs)

A memoized (aka pure) component does an additional thing which a non-memoized Component doesn't do: shallowly compare props. But that's cheap enough and seems worth it, compared to (re)iterating the bundles.

Wrt the ReactLocalization class: I already had a comment typed out protesting against it, but then I gave myself a real think about it and ended up agreeing and bringing it back. Apparently I also removed ReactLocalization's comments in the process without finding a better place for them. Shame on me!
For posterity reasons why I think bringing it back is a good idea:
- JS engines do a better job at optimizing for classes with known methods and fields
- the concerns of the Provider and the Localization are separate enough
- context is initialized with the same type as it's being used with (which I failed to do before)

And for completeness-sake a reason against it would've been that it moves related code further apart, making it potentially a little harder to grok and edit.

Thanks for the thorough review!